accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/PrefetchingUserManager.java

```diff
@@ -0,0 +1,196 @@
+package biz.netcentric.cq.tools.actool.authorizableinstaller.impl;
+
+import static biz.netcentric.cq.tools.actool.history.PersistableInstallationLogger.msHumanReadable;
+
+import java.security.Principal;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jcr.RepositoryException;
+import javax.jcr.ValueFactory;
+
+import org.apache.jackrabbit.JcrConstants;
+import org.apache.jackrabbit.api.security.user.Authorizable;
+import org.apache.jackrabbit.api.security.user.Group;
+import org.apache.jackrabbit.api.security.user.Query;
+import org.apache.jackrabbit.api.security.user.QueryBuilder;
+import org.apache.jackrabbit.api.security.user.User;
+import org.apache.jackrabbit.api.security.user.UserManager;
+import org.apache.jackrabbit.oak.spi.security.user.UserConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import biz.netcentric.cq.tools.actool.helper.Constants;
+import biz.netcentric.cq.tools.actool.history.InstallationLogger;
+
+/**
+ * <p>
+ * Special user manager that prefetches
+ * <ul>
+ * <li>all groups, system users and anonymous (but not regular users)</li>
+ * <li>all memberships between the preloaded authorizables</li>
+ * <ul>
+ * upon creation. 
+ * </p>
+ * <p>
+ * The membership relationships are cached in lookup maps to quickly be able to query the repository state for both
+ * authorizable.declaredMemberOf() (this we call in this class also) and group.getDeclaredMembers(). Having called declaredMemberOf() for
+ * all groups of the repository has also retrieved all memberships (expect regular user memberships), hence the method getDeclaredMembers()
+ * does not have to be called anymore. This is particularly useful because that way the AC tool does not have to iterate over a potentially
+ * huge number of users in production to then only filter out a few relevant groups in the end (the AC Tool does not touch user memberships
+ * to groups).
+ * </p>
+ */
+class PrefetchingUserManager implements UserManager {
```

Comment:
UserManager is a provider-type (https://github.com/apache/jackrabbit-oak/blob/trunk/oak-jackrabbit-api/src/main/java/org/apache/jackrabbit/api/security/user/UserManager.java), therefore this implementation is prone to break with future upgrades....

Since Jackrabbit/Oak doesn't provide a delegate class I would rather put this logic in a dedicated class which does not implement UserManager and only implement those methods which are actually called.

Reply:
Fair enough, see https://github.com/Netcentric/accesscontroltool/blob/3b967710b9b09979b0c4bfdee41123f7175da0d0/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/authorizableinstaller/impl/AuthInstallerUserManager.java#L13 to ensure new methods in UserManager will not break the AC tool in future